Logic rules dependency graph

viktorvanwijk opened the pull request:
Partly closes #2409

~~Dependent on #5936~~

[skip: e2e]

**Changes**

* Added routines to build and analyze a dependency graph
* Added analysis of logic rules when saving them (behind a form feature flag)
* Added fetching which logic rules to execute from the step (behind a form feature flag)

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Dockerfile/scripts

  - [x] Updated the Dockerfile with the necessary scripts from the `./bin` folder

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

viktorvanwijk committed:
:heavy_plus_sign: [#2409] Add networkx dependency

For performing (logic) graph analysis.

viktorvanwijk committed:
:sparkles: [#2409] Add routines to build and analyze a logic rule dependency graph

To ensure that logic rules are executed in the correct order and result in a single state,
the dependency tree of rules must be a directed acyclic graph.

We can build this graph using the determined input and output variables of the logic rules.
Performing a (topological) sort on this graph will ensure that rules with input variables which are
output variables of other rules, are executed first.

If we were not able to determine a step from the input or output variables, for example when both
the input and output variable is user defined, we need to analyze the graph and manually assign
steps to the rule based on the predecessors.

Notes that this can also be achieved by processing the graph breadth first. networkx has routines
that can do this, but that only works by passing in a starting node. Because the graph is not guaranteed
to be connected, we'd have to do some additional analysis of starting nodes. Sorting was chosen for
implementation simplicity, as far as simplicity goes here :grimacing:

viktorvanwijk committed:
:sparkles: [#2409] Add form steps field to form logic

This will be used to persist the form steps on which to execute each rule.

viktorvanwijk committed:
:sparkles: [#2409] Ensure logic rules are analyzed upon saving

I considered adding a separate validator for this, but the context of the serializer is updated, so I thought it might create difficult to follow flows.

Note that this validation is performed on form save, which might lead to out-of-sync analysis if a form designer were to modify form definitions and/or
logic rules outside the form edit page. Removing components from a form definition will result in broken rules anyway, even without this analysis. But
if an individual logic rule was edited, it would require the form to be saved again right now, which is not ideal.

According to support, no one has ever complained or asked questions about modifying individual logic rules in the admin, so for now we will assume that
it doesn't happen and put a disclaimer in the changelog about it.

viktorvanwijk committed:
:recycle: [#2409] Assign all steps of input variables

If we cannot resolve a step from the output variables, we look at the input variables of each rule/action.
Previously, we assigned only the last step to ensure we have all data available, and avoids unnecessarily
executing the rule on multiple steps.

However, this is a change in behaviour from the current situation, where we would execute a rule on all
steps (unless a "trigger_from_step" was defined). So to ensure backwards compatibility, we assign input
variable steps in these cases.

In the future this could be revisited to further optimize rule assignment.

viktorvanwijk committed:
:triangular_flag_on_post: [#2409] Add feature flag for logic rule analysis

For now, we want to keep supporting the current way of executing logic rules, so we add a form feature flag with default opt-out.
This gives municipalities the time to test and adjust their forms if needed.

viktorvanwijk committed:
:zap: [#2409] Ensure logic rules are fetched from the form step

Because the rules and steps are now related, we can retrieve them from the step, and avoid executing all rules unrelated to the current step.

We support both old and new behaviour based on the related form feature flag.

viktorvanwijk committed:
:label: [#2409] Ensure new files are type checked

sergei-maertens commented on src/openforms/forms/models/form.py:
```diff
@@ -397,6 +397,16 @@ class Form(models.Model):
+            "Each rule will be automatically assigned to one or more steps on which "
+            "it will be executed."
+        ),
+    )
```
should we widen the scope of this feature flag to opt into the new logic evaluation entirely? Or, do we:

* expect that we cannot gate the client-side logic evaluation behind a feature flag (initially) or
* expect that we need to decouple rule analysis from client-side logic evaluation?

If possible, I'd like to avoid a situation where two different feature flags need to be setup to use the new logic evaluation, or misconfigurations are possible where the feature flags are mutually exclusive and require validation to prevent that. If we can hang up the entire "new logic evaluation for performance" thing on a single feature flag, that'd be ideal, but then I'd probably suggest renaming this feature flag to something that covers the scope better.

Of course, this all depends on how far we can go with the feature flags, which I can't really tell yet :)

viktorvanwijk replied:
Yeah my intention was definitely a single feature flag for all the changes. I will rename it :)

sergei-maertens commented on src/openforms/forms/api/serializers/logic/form_logic.py:
```diff
@@ -21,6 +30,72 @@ class FormLogicListSerializer(ListWithChildSerializer):
+        cycles = detect_cycles(graph)
+        if cycles:
+            msg = _("Rule contains cycles through variable(s): {variables}.")
+            errors = defaultdict(list)
```
```suggestion
            errors: defaultdict[str, list[ErrorDetail]] = defaultdict(list)
```

sergei-maertens commented on src/openforms/forms/api/serializers/logic/form_logic.py:
```diff
@@ -21,6 +30,72 @@ class FormLogicListSerializer(ListWithChildSerializer):
+        new_rule_order = resolve_order(graph)
+
+        # Reorder the incoming data according to the determined order.
+        data_new, steps = [], []
```
type hints please 😇

sergei-maertens commented on src/openforms/forms/tests/test_api_form_logic_bulk.py:
```diff
@@ -1519,3 +1519,227 @@ def test_create_rule_with_evaluate_dmn_action_empty_config(self):
+            )
+
+        with self.subTest("Rules of step 2"):
+            self.assertEqual(list(step_2.logic_rules.all()), [form_logic_qs[2]])
```
here too I think `self.assertQuerySetEqual` would be nice

sergei-maertens reviewed (changes requested):
Nice work on the analysis and graph theory application!

The bulk of the suggestions aim to improve the readability of the code.

Questions and edge cases:

* logic rule analysis triggers on form save. What happens when someone edits a logic rule or form definition in the admin outside of the regular form-creation-form page?
    * form definitions can remove components -> they remove variables that may be used in logic rules
    * logic rule edits would result in an out-of-sync analysis result compared to the actual rule and lead to weird evaluation bugs

viktorvanwijk commented on src/openforms/forms/tests/test_logic_analysis.py:
```diff
@@ -0,0 +1,627 @@
+            assert cycles is not None
+            self.assertEqual(len(cycles), 1)
+            cycle = list(cycles)[0]
+            self.assertCountEqual(set(cycle["rules"]), {rule_1, rule_2, rule_3})
```
Ah, yeah, I think I had a list comparison before

viktorvanwijk commented on src/openforms/forms/logic_analysis.py:
```diff
@@ -0,0 +1,139 @@
+from .models import FormLogic, FormStep
+
+
+def create_graph(rules: Iterable[FormLogic]) -> DiGraph:
```
Yeah this is what I wanted to do as well, but it raises a `TypeError: type 'DiGraph' is not subscriptable`

viktorvanwijk commented on src/openforms/forms/logic_analysis.py:
```diff
@@ -0,0 +1,139 @@
+
+class Cycle(TypedDict):
+    rules: Collection[FormLogic]
+    variables: Collection[str]
```
I think I've seen too many type hints for the components :sweat_smile:

viktorvanwijk commented on src/openforms/forms/models/logic.py:
```diff
@@ -25,6 +25,13 @@ class FormLogic(OrderedModel):
+    form_steps = models.ManyToManyField(
+        FormStep,
+        verbose_name=_("form steps"),
+        help_text=_("List of form steps on which this logic rule should be executed."),
```
No for the form steps it doesn't really matter, was just wondering :). Will change it to "Collection"

viktorvanwijk committed:
:ok_hand: [#2409] Process PR feedback

- Add type hints
- Change `Cycle` to dataclass because it's easier to work with
- Added comments and update docstring for clarification
- Added reminder to changelog about logic rule analysis being out-of-sync

Note that it was considered to not pass a `pk` when validating the logic rules, and instead use a custom hash override on
the `FormLogic` model to ensure they can be used in the graph. There is a potential risk of someone calling `.save` on
those rules, and overriding existing rules in the DB. But with the comment and renamed variable, it _should_ be sufficiently
clear enough that these rules are just temporary and should live in-memory only.
